test/app-luatest/console_debugger_session_test.lua

```diff
@@ -298,25 +298,38 @@ local function run_debug_session(cmdline, sequence, tmpfile, header)
                 debuglog('Execute command: "'..trim(cmd)..'"')
             end
 
-            local result
-            local clean_cmd = trim(cmd)
-            -- there should be empty stderr - check it before stdout
-            local errout = fh:read({ timeout = 0.05, stderr = true})
-            debuglog('stderr output: ', trim(errout))
-            if first and errout then
-                -- we do not expect anything on stderr
-                -- with exception of initial debugger header
-                t.assert_str_contains(trim(errout), header, false)
+            if first then
+                -- only initial debugger header is expected on stderr
+                t.helpers.retrying({timeout = 3, delay = 0.05},
+                function(expected)
+                    local errout = trim(unescape(fh:read({ timeout = 0.5,
+                                                           stderr = true })))
+                    t.assert_str_contains(errout, expected, false)
+                    debuglog('stderr output: ', trim(errout))
+                end, header)
                 first = false
             else
+                -- stderr should be empty, check it before reading stdout
+                local errout = fh:read({ timeout = 0.05, stderr = true })
+                if errout == nil or trim(errout) == '' then
+                    debuglog('stderr output: ', trim(errout))
+                end
                 t.assert(errout == nil or trim(errout) == '')
             end
-            repeat
-                result = trim(unescape(fh:read({ timeout = 0.5 })))
-                debuglog('stdout output:', result)
-            until result ~= '' or result ~= clean_cmd
-            if expected ~= '' then
-                t.assert_str_contains(result, expected, false)
+
+            -- The output is returned in several chunks, first the prompt,
+            -- then the command, and finally the result. In most cases it
+            -- is can be read in one go, but to prevent flacking the reading
+            -- may be retried.
+            if (expected ~= '') then
+                t.helpers.retrying({timeout = 3, delay = 0.05},
+                function(expected)
+                    local result = trim(unescape(fh:read({ timeout = 0.5 })))
+                    t.assert_str_contains(result, expected, false)
+                    debuglog('stdout output:', result)
+                end, expected)
```

Comment:
It doesn't solve the problem from the comment above:

> The output is returned in several chunks

For example, you're looking for `foo`. The first chunk is `fo`, the second is `o`. This code tries to find `foo` in `fo` and `o` instead of looking into the concatenated result.

Reply:
The comment explains what is ment by 'chunk' - one of the followup: prompt, cmd, result.

Reply:
(I don't mind concatenating the results, but it doesn't seem necessary)

Reply:
This is the behavior that you observe on some particular setup. It is not guaranteed by `read()`. I would prefer to lean on guarantees, especially when the work is about a test case stabilization.

If you're interested, whether it matters in practice: yes, it does. I've observed reading from a pipe with randomly splitted chunks during work on the interactive_tarantool.lua helper.